### ISSUE TITLE
Several "global" stores and a local store. Proper way?

### ISSUE THREAD

**Issue by nbrady-techempower** [mobxjs/mobx-react#156]
The current structure I have seems to be working correctly *most of the time* but I'm having issues that make me wonder if I'm structuring things properly. The project is using ES6/ES7 syntax.

I have a few mobx global stores that I'm passing down through a provider. These are stores that a large chunk of my application needs to know about. Then I have "local" stores to manage the state of each component. However, in those local stores, I need access to the global stores that I'm injecting. So here's an example of what I'm doing.

```javascript
import React from 'react';
import { observer } from 'mobx-react';

import HeaderBlockStore from './HeaderBlockStore';

@observer(['AppStore','ViewsStore'])
class HeaderBlock extends React.Component {

  constructor(props) {
    super(props);
  }

  componentWillMount() {
    this.store = new HeaderBlockStore(this.props);
  }

/** more code **/
```

So I'm importing the local store and creating a new store once this `HeaderBlock` component is created. I'm passing in the props of this component to its store so I can access the global `AppStore` and `ViewsStore` from within the `HeaderBlockStore`.

Basically, I have a layout that looks something like:
```
<Header>
<View>
```

And each `<View>` can manipulate the `ViewsStore`. `HeaderBlockStore` actions require knowledge of `ViewsStore`. (Think links in the header changing their actions based on what view we're currently on.) 

However, it seems that when ViewsStore's observables change, the corresponding props don't always change. I'm curious about the proper way to manage this problem (handle global stores in a more localized store to manipulate actions) as I don't see a clear answer in any documentation.

**Comment by mweststrate** [comment]
can you elaborate on "ViewsStore's observables change, the corresponding props don't always change"? What are "corresponding props"? Can you share some simple code maybe?

**Comment by nbrady-techempower** [comment]
@mweststrate The observables in ViewsStore that are being injected into HeaderBlock aren't updating when I change them via some action in ViewsStore. At first glance does the pattern I'm using make sense?

https://github.com/everlib/everpaste/tree/eacdadab12f514ace73c867056210f4e457c5361/src

This is the state I'm currently at. There might be a thing or two out of place, like I have `componentWillMount`, `componentWillUpdate` and `componentWillReact` in HeaderBlock because I'm just trying a number of different things to figure out what's happening. The ViewsStore is in `/src/stores` and the other "local" stores are with their components in `/src/components/blocks/` and `/src/components/views/`.

**Comment by mweststrate** [comment]
do you have local stores that are passed down, and then stored in local reference the child components? (e.g. some thing like `this.myBeaufifulStore= this.props.BeautifulStore` in `constructor` or `componentWillMount`? That might play tricks with you even new props are passed down, but `this.myBeautifulStore` isn't updated accordingly

**Comment by mweststrate** [comment]
I think `componentWillReceiveProps` is the proper hook here: https://github.com/everlib/everpaste/blob/eacdadab12f514ace73c867056210f4e457c5361/src/components/blocks/HeaderBlock/HeaderBlock.jsx#L20

Did you check https://mobxjs.github.io/mobx/best/react.html btw? Is your project locally runable?

**Comment by mweststrate** [comment]
Ah and double check, are all components that render observables marked with `observer`? Besides that general setup seems fine

**Comment by mweststrate** [comment]
btw could you try mobx-react@4.0.0-beta.2 ?

**Comment by nbrady-techempower** [comment]
Yes on all counts. I've read through that section of the docs. Fixed the lifecycle hook. Tried the beta candidate. And it looks like everything that should have `observer` does. The curious thing is, the console displays this error message:

```
HeaderBlock.jsx:204 Uncaught TypeError: _this2.props.ViewsStore.current.saveButton is not a function(…)
```

Yet it will still execute. I'm wondering if it had something to do with the way I was handling the `current` observable object in ViewsStore. The project is locally runable - babel, webpack, etc - are all being run from the local modules folder. Nothing global but npm & node 6.

**Comment by mweststrate** [comment]
The error seems correct? https://github.com/everlib/everpaste/blob/eacdadab12f514ace73c867056210f4e457c5361/src/components/views/Views.store.js#L31 Where / how do you update `current` ?

**Comment by nbrady-techempower** [comment]
https://github.com/everlib/everpaste/blob/eacdadab12f514ace73c867056210f4e457c5361/src/components/views/PasteView/PasteViewStore.js#L13-L16

In the store of the first component that's rendered. PasteView is the sibling of HeaderBlock.

**Comment by nbrady-techempower** [comment]
Is what I'm expecting incorrect? I'm expecting that when the PasteViewStore is constructed, it will update ViewsStore and the HeaderBlock link will now hold the proper function.

**Comment by mweststrate** [comment]
Depending on the rendering order of components is a very error prone abstraction to make, I would be careful with that, and inverse the reasoning: first construct your stores etc based on the new state, and then render. That makes testing etc also a lot easier.

But for now you could start by simply checking is the the constructor of the `PasteViewStore` run before the one from HeaderBlock? And is the initializer (that creates the `saveButton`) run before the constructor of that store (I think so, from the top of my head). 

Finally note that `observable({ x: function() {} })` actually creates a computed value `x`, not an action. (Which is stupid legacy is being deprecated). So you might want to use `asReference` to be sure:

Also note that the assignment to `current.text` is a no-op, because in the next line you throw away `current` itself

```
    this.props.ViewsStore.current.text : '';

    this.props.ViewsStore.current = observable({
      currentView: 'PasteView',
      saveButton: asReference(this.saveButton)
    });
```

**Comment by nbrady-techempower** [comment]
Oh, yeah that line ` this.props.ViewsStore.current.text : '';` is part of a ternary operation

```
    this.defaultTxt = this.props.pathname == '/edit' ?
      this.props.ViewsStore.current.text : '';
```

I guess the whole idea of constructing stores based on the new state is where the whole thing gets convoluted for me if you have multiple local stores for sibling components that all share a global store. The Header is constructed, passes it's props into HeaderStore. HeaderStore changes the global ViewsStore, updates Header but maybe not before PasteView is constructed. PasteView changes it's local store, PasteViewStore, which can alter the global ViewsStore. Which then alters HeaderStore and... whaaat?

With your explanation the component that holds the siblings Header and PasteView would compute all of their stores and the global ViewsStore before rendering them? I'm not quite sure how to approach that, especially when the component to be displayed after <Header /> is unknown and being passed into the layout based on the route. The only solution that makes sense is just use one global store and forego local stores, but that's not scalable. However, I feel like that's what most of the docs are showing. 

The single `asReference` change solved my issue here. I have a bit of rewinding to do probably to make sure it's not that in addition to something else, but I think that takes care of it. Did I miss this in the documentation or is it a product of the beta update I did? Still I know there is probably unnecessary rendering going on because of my current design.

In any event, these are not really mobx issues anymore, since that solved it. Thanks for your help. I feel like I'll be using mobx for quite some time, so hopefully I can make a contribution to the project soon. That's the best way to repay people like you for their help. Thanks again!

**Comment by mweststrate** [comment]
You might find this an interesting read: https://medium.com/@mweststrate/how-to-decouple-state-and-ui-a-k-a-you-dont-need-componentwillmount-cc90b787aa37

I think your current setup is quite sensitive changes in your component tree, as that would influence the order in which your store methods are invoked. For routing you are not per se dependent on having components mounted. I think that is a conceptual draw back of react-router: it makes state dependent on the mounting of components, so it kinda breaks view-as-function-of-state principle. You could for example take a look at [mobx-router](https://github.com/kitze/mobx-router)

Another approach you could take is using promise or a promise like model: register callbacks that will be invoked as soon as the necessary data is available. I think both approaches can suffice to decouple the order in which your logic is executed from the order in which your components are rendered. I hope that helps!

**Comment by nbrady-techempower** [comment]
Helps a ton @mweststrate. Thanks again for this and all the open-source work you guys do!
